c2_utils/src/mfx_defs.cpp

```diff
@@ -42,57 +42,181 @@ static void InitMfxFrameHeader(
+    uint32_t nOHeight = input_info.Height;
+    uint32_t nOPitch = stride;
+
+    uint32_t nCropX = srf->Info.CropX;
+    uint32_t nCropY = srf->Info.CropY;
+    uint32_t nCropW = srf->Info.CropW;
+    uint32_t nCropH = srf->Info.CropH;
+    uint32_t nPitch = srf->Data.Pitch;
+
+    if (!MFX_C2_IS_COPY_NEEDED(srf->Data.MemType, input_info, srf->Info)) {
+        srf->Data.Y  = data;
+        srf->Data.UV = data + nOWidth * nOHeight;
+        srf->Data.Pitch = nOPitch;
+    } else {
+        uint32_t i = 0;
+        uint8_t* Y  = data;
+        uint8_t* UV = data + nOPitch * nOHeight;
+
+        // if input surface width or height is not 16bit aligned, do copy here
+        for (i = 0; i < nCropH/2; ++i)
+        {
+            // copying Y
+            uint8_t *src = Y + nCropX + (nCropY + i)*nOPitch;
+            uint8_t *dst = srf->Data.Y + nCropX + (nCropY + i)*nPitch;
+            std::copy(src, src + nCropW, dst);
+
+            // copying UV
+            src = UV + nCropX + (nCropY/2 + i)*nOPitch;
+            dst = srf->Data.UV + nCropX + (nCropY/2 + i)*nPitch;
+            std::copy(src, src + nCropW, dst);
+        }
+        for (i = nCropH/2; i < nCropH; ++i)
+        {
+            // copying Y (remained data)
+            uint8_t *src = Y + nCropX + (nCropY + i)*nOPitch;
+            uint8_t *dst = srf->Data.Y + nCropX + (nCropY + i)*nPitch;
+            std::copy(src, src + nCropW, dst);
+        }
+    }
+
+    return res;
 }
 
-void InitMfxFrameHW(
+mfxStatus InitMfxFrameHW(
     uint64_t timestamp, uint64_t frame_index,
     mfxMemId mem_id,
     uint32_t width, uint32_t height, uint32_t fourcc, const mfxFrameInfo& info, mfxFrameSurface1* mfx_frame)
 {
     MFX_DEBUG_TRACE_FUNC;
+    mfxStatus res = MFX_ERR_NONE;
 
     InitMfxFrameHeader(timestamp, frame_index, width, height, fourcc, info, mfx_frame);
 
     MFX_ZERO_MEMORY(mfx_frame->Data);
     mfx_frame->Data.MemType = MFX_MEMTYPE_EXTERNAL_FRAME;
     mfx_frame->Data.MemId = mem_id;
+
+    return res;
+}
+
+uint32_t MFXGetSurfaceSize(uint32_t FourCC, uint32_t width, uint32_t height)
+{
+    MFX_DEBUG_TRACE_FUNC;
+
+    mfxU32 nbytes = 0;
+
+    switch (FourCC) {
+        case MFX_FOURCC_I420:
+        case MFX_FOURCC_NV12:
+            nbytes = width * height + (width >> 1) * (height >> 1) + (width >> 1) * (height >> 1);
+            break;
+        case MFX_FOURCC_I010:
+        case MFX_FOURCC_P010:
+            nbytes = width * height + (width >> 1) * (height >> 1) + (width >> 1) * (height >> 1);
+            nbytes *= 2;
+            break;
+        case MFX_FOURCC_RGB4:
+            nbytes = width * height * 4;
+            break;
+        default:
+            break;
+    }
+
+    return nbytes;
+}
+
+uint32_t MFXGetFreeSurfaceIdx(mfxFrameSurface1 *SurfacesPool, uint32_t nPoolSize)
+{
+    MFX_DEBUG_TRACE_FUNC;
+
+    for (uint32_t i = 0; i < nPoolSize; i++) {
+        if (0 == SurfacesPool[i].Data.Locked)
+            return i;
+    }
```

Comment:
this design not very flexible, this is system memory so there is no requirenment from MSDK to pre-allocate all the buffers bofore the encoder porocess start,  then we can just preallocte serverl surfaces during init status and allocate more during encode process if more surface is required?
With fixed decode surface, maybe not so much surface needed, another case more surface may needed.

Reply:
That's needed for input frames which is not 16byte aligned. Both encoder and decoder need the surface  to meet 16 byte alignment requiremnt as encoding is processed one Macroblock by one Macroblock (16x16).